xdsl/dialects/pdl.py

```diff
@@ -0,0 +1,323 @@
+
+from xdsl.dialects.builtin import (ArrayAttr, IntegerAttr, IntegerType,
+                                   StringAttr)
+from xdsl.ir import (Attribute, Dialect, MLIRType, OpResult, Operation,
+                     ParametrizedAttribute, Region, SSAValue)
+from xdsl.irdl import (AttrSizedOperandSegments, OpAttr, Operand, OptOpAttr,
+                       OptOperand, OptRegion, ParameterDef, VarOpResult,
+                       VarOperand, irdl_attr_definition, irdl_op_definition)
+from xdsl.utils.exceptions import VerifyException
+
+
+@irdl_attr_definition
+class AttributeType(ParametrizedAttribute, MLIRType):
+    name = "pdl.attribute"
+
+
+@irdl_attr_definition
+class OperationType(ParametrizedAttribute, MLIRType):
+    name = "pdl.operation"
+
+
+@irdl_attr_definition
+class TypeType(ParametrizedAttribute, MLIRType):
+    name = "pdl.type"
+
+
+@irdl_attr_definition
+class ValueType(ParametrizedAttribute, MLIRType):
+    name = "pdl.value"
+
+
+AnyPDLType = AttributeType | OperationType | TypeType | ValueType
+
+_RangeT = TypeVar("_RangeT",
+                  bound=AttributeType | OperationType | TypeType | ValueType,
+                  covariant=True)
+
+
+@irdl_attr_definition
+class RangeType(Generic[_RangeT], ParametrizedAttribute, MLIRType):
+    name = "pdl.range"
+    element_type: ParameterDef[_RangeT]
+
+
+@irdl_op_definition
+class ApplyNativeConstraintOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdlapply_native_constraint-mlirpdlapplynativeconstraintop
+    """
+    name: str = "pdl.apply_native_constraint"
+    # https://github.com/xdslproject/xdsl/issues/98
+    # name: OpAttr[StringAttr]
+    args: Annotated[VarOperand, AnyPDLType]
+
+    def verify_(self) -> None:
+        if 'name' not in self.attributes:
+            raise Exception(
+                "ApplyNativeConstraintOp requires a 'name' attribute")
+
+        if not isinstance(self.attributes['name'], StringAttr):
+            raise Exception("expected 'name' attribute to be a StringAttr")
+
+    @staticmethod
+    def get(name: str, args: list[SSAValue]) -> ApplyNativeConstraintOp:
+        return ApplyNativeConstraintOp.build(
+            result_types=[],
+            operands=[args],
+            attributes={"name": StringAttr(name)})
+
+
+@irdl_op_definition
+class ApplyNativeRewriteOp(Operation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLOps/#pdlapply_native_rewrite-mlirpdlapplynativerewriteop
+    """
+    name: str = "pdl.apply_native_rewrite"
+    # https://github.com/xdslproject/xdsl/issues/98
+    # name: OpAttr[StringAttr]
+    args: Annotated[VarOperand, AnyPDLType]
+    res: Annotated[VarOpResult, AnyPDLType]
+
+    def verify_(self) -> None:
+        if 'name' not in self.attributes:
+            raise Exception("ApplyNativeRewriteOp requires a 'name' attribute")
+
+        if not isinstance(self.attributes['name'], StringAttr):
+            raise Exception("expected 'name' attribute to be a StringAttr")
```

Comment:
We can't declare it as a unual attribute because it shadows `name` from `Operation`.
I'll try to come up with a good workaround, but it will be heavily breaking.